### ISSUE TITLE
Questions of calculateContactBasis() of The contacts.cpp

### ISSUE THREAD

**Issue by lch32111** [idmillington/cyclone-physics#49]
The method calculateContactBasis() of The contacts.cpp has a purpose of calculating orthonormal basis in contact coordinates.

So, I read the contents that the contact normal will be used as a X-axis in the contact coordinates.
To calculate the Z-axis and Y-axis in the contact coordinates, the code uses cross product first between X-axis(contact normal) and World-Up(0,1,0) Vector. and the result is Z-axis.

With the result, it calculates the Y-axis, using the cross product between Z-axis and X-axis.

So, the preliminary code of makeOrthonormalBasis in the book is like:

```
void makeOrthonormalBasis(const Vector& x, Vector* y, Vector* z)
{
       // Calculate z from the vector product of x and y.
       z = (*x) % (*y);

       // Check for y and z in parallel
       if(z->squaredMagnitude() == 0.0) return;

       // Calculate y from the vector product of z and x
       (*y) = (*z) % x;

      // Normalize the output vectors
       y->normalize();
       z->normalize();
}
```

I think this method is right and the calculateContactBasis() is based on this logic.

```
        // Scaling factor to ensure the results are normalised
        const real s = (real)1.0f/real_sqrt(contactNormal.z*contactNormal.z +
            contactNormal.x*contactNormal.x);

        // The new X-axis is at right angles to the world Y-axis
        contactTangent[0].x = contactNormal.z*s;
        contactTangent[0].y = 0;
        contactTangent[0].z = -contactNormal.x*s;

        // The new Y-axis is at right angles to the new X- and Z- axes
        contactTangent[1].x = contactNormal.y*contactTangent[0].x;
        contactTangent[1].y = contactNormal.z*contactTangent[0].x -
            contactNormal.x*contactTangent[0].z;
        contactTangent[1].z = -contactNormal.y*contactTangent[0].x;
```
And My questions start here.

Because

```
Z-axis = crossProduct(X-axis, Y-axis)

if X-axis = (x1, x2, x3), Y-axis = (y1, y2, y3) = (0, 1, 0)

crossProduct(X-axis, Y-axis) = 

|  i    j    k |
| x1 x2 x3 |
| 0   1   0  |

= (x2 * 0 - x3 * 1, x1 * 0 - x3 * 0, x1 * 1 - x2 * 0)
= (-x3, 0, x1)
```
But the code is..
```
contactTangent[0].x = contactNormal.z*s;
contactTangent[0].y = 0;
contactTangent[0].z = -contactNormal.x*s;
```
When we just write the cross product between Z-axis and X-axis
```
Y-axis = crossProduct(Z-axis, X-axis)

if Z-axis = (z1, z2, z3) = (z1, 0, z3) (the calculated one from the cross product between X and Y axis, 
   X-axis = (x1, x2, x3)

crossProduct(Z-axis, X-axis) = 

|  i    j    k |
| z1 0 z3 |
| x1 x2 x3 |

= (0 * x3 - z3 * x2, -(z1 * x3 - z3 * x1), z1 * x2 - 0 * x1 )
= (-z3 * x2, - z1 * x3 + z3 * x1, z1 * x2)
```

We get the above equation. However, the code of calculating Y-axis is

```
contactTangent[1].x = contactNormal.y*contactTangent[0].x;
contactTangent[1].y = contactNormal.z*contactTangent[0].x - contactNormal.x*contactTangent[0].z;
contactTangent[1].z = -contactNormal.y*contactTangent[0].x;
```

The code should be like this
```
contactTangent[1].x = - contactNormal.y * contactTangent[0].z;
contactTangent[1].y = -contactNormal.z * contactTangent[0].x + contactNormal.x * contactTangent[0].z;
contactTangent[1].z = contactNormal.y * contactTangent[0].x;
```

In summary, the questions are:

1. are the lines of calculating the orthonormal basis right?

**Comment by e2-dunstan** [comment]
Hello, did you figure this out? I have a feeling that a bug in my code is to do with this.

**Comment by lch32111** [comment]
@e2-dunstan  No.

I actually gave up clinging to this code.
After this book, I changed my physics engine base, box2D.
If you try to implement your own physics engine, I recommend you see box2D presentations on GDC. It's dealing with jitter problems in modern way. 

But Game Physics Engine Developemtn Book is not handling this problem well.
And the sample code is not maintained now.

see this link too : https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=6&t=63
BulletPhysics, and Box2d are famous physics engine which you should study from.

Good luck.
